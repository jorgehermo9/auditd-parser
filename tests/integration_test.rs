use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
