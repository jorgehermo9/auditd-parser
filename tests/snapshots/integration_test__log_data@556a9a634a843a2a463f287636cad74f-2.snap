---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/go-libaudit/test2.log:4"
expression: result
info: "type=PATH msg=audit(1170021493.977:283): item=0 name=\"maildrop\" inode=14911367 dev=03:07 mode=040730 ouid=890 ogid=891 rdev=00:00 obj=system_u:object_r:postfix_spool_maildrop_t:s0"
---
{
  "record_type": "PATH",
  "timestamp": 1170021493977,
  "id": 283,
  "fields": {
    "dev": "03:07",
    "inode": "14911367",
    "item": "0",
    "mode": {
      "attributes": [],
      "file_type": "directory",
      "group": [
        "write",
        "exec"
      ],
      "other": [],
      "user": [
        "read",
        "write",
        "exec"
      ]
    },
    "name": "maildrop",
    "obj": {
      "level": {
        "sensitivity": "s0"
      },
      "module": "SELinux",
      "role": "object_r",
      "type": "postfix_spool_maildrop_t",
      "user": "system_u"
    },
    "ogid": 891,
    "ouid": 890,
    "rdev": "00:00"
  }
}
